File: examples/larger_topo.cpp


#include <iostream>
//#include <hwloc.h>

#include "sys-sage.hpp"

using namespace sys_sage;

using std::cout;
using std::endl;

int main(int argc, char *argv[])
{
    //create root Topology and one node
    Topology* topo = new Topology();
    int tot_nodes=8;
    Node* n = new Node(1);
    n->SetParent((Component*)topo);
    topo->InsertChild((Component*)n);

    cout << "create topology..." << endl;
    std::string xmlPath = EXAMPLE_DIR + std::string("/skylake_hwloc.xml");
    std::string bwPath = EXAMPLE_DIR + std::string("/skylake_caps_numa_benchmark.csv");

    for(int n_idx=0; n_idx<tot_nodes; n_idx++)
    {
        Node* n = new Node(n_idx);
        n->SetParent((Component*)topo);
        topo->InsertChild((Component*)n);
        if(parseHwlocOutput(n, xmlPath) != 0)
        {
            cout << "error parsing hwloc in path " << xmlPath << endl;
            return 1;
        }
        if(parseCapsNumaBenchmark((Component*)n, bwPath, ";") != 0)
        {
            cout << "failed parsing caps-numa-benchmark in path " << bwPath << endl;
            return 1;
        }
    }

    std::string output_name = "sys-sage_sample_output.xml";
    cout << "Exporting as XML to " << output_name << endl;
    exportToXml(topo, output_name);

    return 0;
}

File: examples/larger_topo.py
import py_sys_sage as pysage 
import sys
import os 

if __name__ == "__main__":
    path_prefix = sys.argv[0]
    path_prefix = os.path.dirname(path_prefix) + "/"
    xmlPath = path_prefix + "example_data/skylake_hwloc.xml"
    bwPath = path_prefix + "example_data/skylake_caps_numa_benchmark.csv"

    # create root Topology and one node
    topo = pysage.Topology()
    n = pysage.Node(topo, 1)
    for n_idx in range(8):
        n = pysage.Node(topo, n_idx)
        if pysage.parseHwlocOutput(n, xmlPath) != 0:
            print("error parsing hwloc in path " + xmlPath)
            sys.exit(1)
        if pysage.parseCapsNumaBenchmark(n, bwPath, ";") != 0:
            print("failed parsing caps-numa-benchmark in path " + bwPath)
            sys.exit(1)

    output_name = "sys-sage_sample_output.xml"
    print("Exporting as XML to " + output_name)
    pysage.exportToXml(topo, output_name)
